x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/StringRuntimeFieldHelper.java

```diff
@@ -0,0 +1,491 @@
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.util.ArrayUtil;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefBuilder;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.ByteRunAutomaton;
+import org.apache.lucene.util.automaton.CompiledAutomaton;
+import org.apache.lucene.util.automaton.RegExp;
+import org.elasticsearch.common.CheckedFunction;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.IntConsumer;
+
+/**
+ * Manages the creation of doc values and queries for {@link String} fields.
+ */
+public final class StringRuntimeFieldHelper {
+    @FunctionalInterface
+    public interface NewLeafLoader {
+        IntConsumer leafLoader(LeafReaderContext ctx, Consumer<String> sync) throws IOException;
+    }
+
+    private final NewLeafLoader newLeafLoader;
+
+    public StringRuntimeFieldHelper(NewLeafLoader newLeafLoader) {
+        this.newLeafLoader = newLeafLoader;
+    }
+
+    public CheckedFunction<LeafReaderContext, SortedBinaryDocValues, IOException> docValues() {
+        return new Values().docValues();
+    }
+
+    public Query existsQuery(String fieldName) {
+        return new Values().new ExistsQuery(fieldName);
+    }
+
+    public Query fuzzyQuery(String fieldName, String value, int maxEdits, int prefixLength, int maxExpansions, boolean transpositions) {
+        return new Values().new FuzzyQuery(fieldName, value, maxEdits, prefixLength, maxExpansions, transpositions);
+    }
+
+    public Query prefixQuery(String fieldName, String value) {
+        return new Values().new PrefixQuery(fieldName, value);
+    }
+
+    public Query rangeQuery(String fieldName, String lowerValue, String upperValue, boolean includeLower, boolean includeUpper) {
+        return new Values().new RangeQuery(fieldName, lowerValue, upperValue, includeLower, includeUpper);
+    }
+
+    public Query regexpQuery(String fieldName, String pattern, int flags, int maxDeterminizedStates) {
+        return new Values().new RegexpQuery(fieldName, pattern, flags, maxDeterminizedStates);
+    }
+
+    public Query termQuery(String fieldName, String value) {
+        return new Values().new TermQuery(fieldName, value);
+    }
+
+    public Query termsQuery(String fieldName, String... values) {
+        return new Values().new TermsQuery(fieldName, values);
+    }
+
+    public Query wildcardQuery(String fieldName, String pattern) {
+        return new Values().new WildcardQuery(fieldName, pattern);
+    }
+
+    private class Values extends AbstractRuntimeValues {
+        private String[] values = new String[1];
```

Comment:
should we just have the collection of values shared between the different types and use a typed ArrayList? Would that be a tragedy in terms of performance?